### PR TITLE
[exporter/datadogexporter] Refactor metrics translation

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

Refactors metrics translation into a new `Translator` struct that bundles reusable components to simplify method calling.
This is a first PR for eventually moving the metrics translation logic into a new separate module. A second PR will follow which separates the Collector-specific logic from the metrics translation bits.

The diff for the entire PR may be a bit difficult to understand; I would recommend checking each individual commit separately (I can split into separate PRs if preferred).

This is an internal refactor only, and should have no visible effects for users of the Datadog exporter.

**Link to tracking Issue:** n/a

**Testing:** Amended unit tests.

### CODE REVIEW

**Review by bogdandrutu — DISMISSED** [review]
Unfortunately we broke the `main` branch, need to do more work to rebase to the correct main

**Comment by mx-psi** [comment]
Got it, I will try to fix this once we can merge to `main` again, thanks for the update!

File: exporter/datadogexporter/internal/translator/metrics_translator.go
Comment (KSerrania) [inline exporter/datadogexporter/internal/translator/metrics_translator.go:45]: I think we should call that field `fallbackHostProvider`, which describes more precisely what that specific `HostnameProvider` is used for in the translator?

File: exporter/datadogexporter/metrics_exporter.go
Comment (KSerrania) [inline exporter/datadogexporter/metrics_exporter.go:41]: Maybe add a comment here to explain the purpose of this line

**Review by KSerrania — APPROVED** [review]
Left a few comments but LGTM overall

File: exporter/datadogexporter/metrics_exporter.go
Comment (mx-psi) [reply exporter/datadogexporter/metrics_exporter.go:41]: It's an ['interface assertion'](https://stackoverflow.com/a/44976881) (there is no official name for it, really). It is a way for the compiler to typecheck that a given struct implements all methods of an interface

**Comment by mx-psi** [comment]
@bogdandrutu PTAL

**Comment by mx-psi** [comment]
Marking as ready to merge since it was approved by a CODEOWNER and it is rebased correctly against main now

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
